cleverhans/attacks_tf.py

```diff
@@ -726,3 +726,126 @@ def compare(x, y):
         # return the best solution found
         o_bestl2 = np.array(o_bestl2)
         return o_bestattack
+
+
+def deepfool_batch(sess, x, pred, logits, grads, X, nb_candidate, overshoot,
+                   max_iter, clip_min, clip_max, nb_classes, feed=None):
+    """
+    Applies the deepfool to a batch of inputs
+    :param sess: TF session
+    :param x: The input placeholder
+    :param pred: The model's symbolic output
+    :param logits: The model's unnormalized output tensor (the input to
+                   the softmax layer)
+    :param grads: Symbolic gradients
+    :param X: Numpy array with sample inputs
+    :param nb_candidate: The number of adversarial class candidates
+    :param overshoot: A termination criterion to prevent vanishing updates
+    :param max_iter: Maximum number of iteration for deepfool
+    :param clip_min: Minimum value for components of the example returned
+    :param clip_max: Maximum value for components of the example returned
+    :param nb_classes: Number of model output classes
+    :return: Adversarial examples
+    """
+
+    X_adv = np.zeros(X.shape)
+    X_adv = deepfool_attack(sess, x, pred, logits, grads, X, nb_candidate,
+                            overshoot, max_iter, clip_min, clip_max, feed=feed)
+
+    return np.asarray(X_adv, dtype=np.float32)
+
+
+def deepfool_attack(sess, x, predictions, logits, grads, sample, nb_candidate,
+                    overshoot, max_iter, clip_min, clip_max, feed=None):
+    """
+    TensorFlow implementation of the deepfool.
+    Paper link: see https://arxiv.org/pdf/1511.04599.pdf
+    :param sess: TF session
+    :param x: The input placeholder
+    :param predictions: The model's symbolic output (linear output,
+        pre-softmax)
+    :param logits: The model's unnormalized output tensor (the input to
+                   the softmax layer)
+    :param grads: Symbolic gradients
+    :param sample: Numpy array with sample input
+    :param nb_candidate: The number of adversarial class candidates
+    :param overshoot: A termination criterion to prevent vanishing updates
+    :param max_iter: Maximum number of iteration for deepfool
+    :param clip_min: Minimum value for components of the example returned
+    :param clip_max: Maximum value for components of the example returned
+    :return: an adversarial sample
+    """
+    import copy
+
+    adv_x = copy.copy(sample)
+    # Initialize the loop variables
+    iteration = 0
+    current = utils_tf.model_argmax(sess, x, logits, adv_x, feed=feed)
+    if current.shape == ():
+        current = np.array([current])
+    w = np.squeeze(np.zeros(sample.shape[1:4]))  # same shape as original image
```

Comment:
Does this imply that `x` should be an image and have the usual TF dimension ordering? If yes, then could you add a check at the beginning of the function that throws an error if the input `x` does not have a compatible shape?